src/JsonApi/Serializer/ItemNormalizer.php

```diff
@@ -0,0 +1,554 @@
+
+            $components['relationships'][] = $relation;
+        }
+
+        return $this->componentsCache[$context['cache_key']] = $components;
+    }
+
+    /**
+     * Populates links and relationships keys.
+     *
+     * @param array       $data
+     * @param object      $object
+     * @param string|null $format
+     * @param array       $context
+     * @param array       $components
+     * @param string      $type
+     *
+     * @return array
+     */
+    private function getPopulatedRelations(
+        $object,
+        string $format = null,
+        array $context,
+        array $components,
+        string $type = 'relationships'
+    ): array {
+        $data = [];
+
+        $identifier = '';
+        foreach ($components[$type] as $relationshipDataArray) {
+            $relationshipName = $relationshipDataArray['name'];
+
+            $attributeValue = $this->getAttributeValue(
+                $object,
+                $relationshipName,
+                $format,
+                $context
+            );
+
+            if ($this->nameConverter) {
+                $relationshipName = $this->nameConverter->normalize($relationshipName);
+            }
+
+            if (!$attributeValue) {
+                continue;
+            }
+
+            $data[$relationshipName] = [
+                'data' => [],
+            ];
+
+            // Many to one relationship
+            if ('one' === $relationshipDataArray['cardinality']) {
+                $data[$relationshipName] = $attributeValue;
+
+                continue;
+            }
+
+            // Many to many relationship
+            foreach ($attributeValue as $attributeValueElement) {
+                if (!isset($attributeValueElement['data'])) {
+                    throw new RuntimeException(sprintf(
+                        'Expected \'data\' attribute in collection for attribute \'%s\'',
+                        $relationshipName
+                    ));
+                }
+
+                $data[$relationshipName]['data'][] = $attributeValueElement['data'];
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * Gets the IRI of the given relation.
+     *
+     * @param array|string $rel
+     *
+     * @return string
+     */
+    private function getRelationIri($rel): string
+    {
+        return $rel['links']['self'] ?? $rel;
+    }
+
+    /**
+     * Gets the cache key to use.
+     *
+     * @param string|null $format
+     * @param array       $context
+     *
+     * @return bool|string
+     */
+    private function getCacheKey(string $format = null, array $context)
+    {
+        try {
+            return md5($format.serialize($context));
+        } catch (\Exception $exception) {
+            // The context cannot be serialized, skip the cache
+            return false;
+        }
+    }
+
+    /**
+     * Denormalizes a resource linkage relation.
+     *
+     * See: http://jsonapi.org/format/#document-resource-object-linkage
```

Comment:
@see and the following @param and @return must be updated ([type] and [description])